clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -8508,6 +8507,16 @@ Sema::BuildExpressionFromDeclTemplateArgument(const TemplateArgument &Arg,
   } else {
     assert(ParamType->isReferenceType() &&
            "unexpected type for decl template argument");
+    if (ParamType->isLValueReferenceType())
+      if (NonTypeTemplateParmDecl *NTTP =
+              dyn_cast_if_present<NonTypeTemplateParmDecl>(TemplateParam)) {
+        QualType TemplateParamType = NTTP->getType();
+        const AutoType *AT = TemplateParamType->getAs<AutoType>();
+        if (AT && AT->isDecltypeAuto())
+          RefExpr = new (getASTContext())
+              ParenExpr(RefExpr.get()->getBeginLoc(),
```

Comment:
Did you try to preserve the references using `SubstNonTypeTemplateParmExpr`s just like what @cor3ntin said? This looks a bit odd to me...

Reply:
Ah, thanks for your remind! `SubstNonTypeTemplateParmExpr` is more suitable here.